wgsl/index.bs

```diff
@@ -6895,21 +6895,76 @@ As such, the same textual location may represent multiple call sites.
     * An [[#address-of-expr|address-of expression]] of a
         [[#var-identifier-expr|variable identifier expression]]
     * A function parameter
-* WGSL assumes no aliasing is present between any combination of
-    function parameters and variables.
-    As such, a function parameter of pointer type must not be used to read or
-    write to any [=memory locations=] of its [=originating variable=] that are
-    also written via:
-    * Another function parameter in the same function
-    * A statement or expression in the function using the originating variable directly
-
-Note: The aliasing restriction applies to memory location written by function
-calls in the function.
 
 Note: Recursion is disallowed because cycles are not permitted among any kinds
 of declarations.
 
-Issue: Revisit aliasing rules for clarity.
+### Aliasing Memory Views ### {#function-aliasing}
+
+[=Memory locations=] can be accessed during the execution of a function using [=memory views=].
+Within a function, each [=memory view=] has a particular root identifier.
+The root identifier can be an [=originating variable=] or a [=formal parameter=]
+of [=pointer type=].
+Locally derived expressions of [=reference type|reference=] or [=pointer
+type|pointer] type may introduce new names for a particular root identifier,
+but each expression has a statically determinable root identifier.
+While the [=originating variable=] of a root identifier is a dynamic concept that
+depends on the [=call sites=] for the function, [SHORTNAME] programs can be
+statically analyzed to determine the set of all possible [=originating
+variables=] for each root identifier.
+When two root identifiers have the same [=originating variable=], they are said to alias.
+It is a [=shader-creation error=] if more than one aliased root identifiers
```

Comment:
There must be some qualification to this.  The accesses have to occur statically the same function call, and dynamically within the same function call instance.

Otherwise, you would not be able to do something like this:

         // p and x are aliased root identifiers.
         fn helper(p:  ptr<function,i32>) {  *p = 2; }
         fn caller() { var x: i32; x = 0; helper(&x); }

or this:

         // double-buffering pattern.
         fn helper2(p:  ptr<function,i32>, r:ptr<function,i32>) {  *p = 2; *r = 3; }
         fn caller() { 
             var x: i32; var y: i32;
             helper2(&x, &y);  // inside helper2, p is aliased to x.
             helper2(&y, &x;   //  inside helper2, r is aliased to x.
         }

Suggest modifying the definition of alias to:

>  Two root identifiers `<dfn>`alias`</dfn>` when they are both in scope and have the same [=originating variable=].

and then:

> It is a shader-creation error if, during one function call, more than one aliased root identifier accesses ...

Reply:
This is upgrading an assumption to a shader-creation error.  If we kept it an assumption then I could approve it without group review.
Suggest landing the editorial change first, then proposing the upgrade to shader-creation error.

Reply:
Fair enough. I realized after the fact I had probably upgraded the error. I've rewritten the conditions as a bullet list.